pilot/cmd/pilot-agent/status/server.go

```diff
@@ -194,6 +223,102 @@ func isRequestFromLocalhost(r *http.Request) bool {
 	return userIP.IsLoopback()
 }
 
+type PrometheusScrapeConfiguration struct {
+	Scrape string `json:"scrape"`
+	Path   string `json:"path"`
+	Port   string `json:"port"`
+}
+
+// handleStats handles prometheus stats scraping. This will scrape envoy metrics, and, if configured,
+// the application metrics and merge them together.
+// The merge here is a simple string concatenation. This works for almost all cases, assuming the application
+// is not exposing the same metrics as Envoy.
+// TODO(https://github.com/istio/istio/issues/22825) expose istio-agent stats here as well
+func (s *Server) handleStats(w http.ResponseWriter, r *http.Request) {
+	var envoy, application []byte
+	var err error
+	if envoy, err = s.scrape(fmt.Sprintf("http://localhost:%d/stats/prometheus", s.envoyStatsPort), r.Header); err != nil {
+		log.Errora(err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if s.prometheus != nil {
+		url := fmt.Sprintf("http://localhost:%s%s", s.prometheus.Port, s.prometheus.Path)
+		if application, err = s.scrape(url, r.Header); err != nil {
+			log.Errora(err)
+			w.WriteHeader(http.StatusInternalServerError)
```

Comment:
@howardjohn I am not sure if it is good idea to return 500 for the metrics endpoint when application metrics can't be fetched. Without metrics merge enabled, if the application has issues to serve HTTP requests, Envoy will keep delivering metrics and users are able to detect any issue when Envoy reports failing requests. With this design, when the application stops responding to requests for some reason, Prometheus will fail to scrape the whole Pod together with Envoy metrics and users will never see 5xx requests coming from this Pod.

Reply:
@cagataygurturk that is a great point. One concern I have if we switch the behavior here is that we may obscure the fact that the application is failing to respond. I have confirmed prometheus will no include the metrics returned in the case of non-200 error codes: https://github.com/prometheus/prometheus/blob/b30654211cf99d244775c25f353d4becdb1aec99/scrape/scrape.go#L560. Would love your feedback here

Reply:
Before writing my initial comment I also thought about your concern. In my opinion, as a user we observe the whole Pod, not specifically the app container. Even though the app container fails to respond to metrics requests, there are still Envoy metrics I would like to see there and in fact thanks to those Envoy metrics I will be able to detect that app container has a serious issue. 

When the whole metric endpoint disappears just because app container's metric endpoint fails, there is a high chance that users are not alerted because the whole metrics endpoint also disappears from Prometheus when a Pod is gracefully replaced during deployments etc. People do not tend to set up alerts for this. In this case we have a risk of keeping a failing Pod in the pool and we will never know that we are sending 5xx's.

In my opinion if app container stops reporting metrics, Envoy metrics should continue to be reported. Users can still use `scrape_samples_scraped` metric to set up alerting policies if they are interested in events where application metrics from a Pod suddenly disappears.

Reply:
We can also add metrics about the scraping with https://github.com/istio/istio/pull/24798. So I think it makes a lot of sense to continue when there are errors. Any opinions @douglas-reid?

Reply:
Thanks for the feedback. Can you take a look at https://github.com/istio/istio/pull/25668?